qupath-gui-fx/src/main/java/qupath/lib/gui/QuPathGUI.java

```diff
@@ -1092,6 +1096,24 @@ else if (version != null) {
 						}
 					}
 				}
+			} else if (e.getButton() == MouseButton.MIDDLE) {
```

Comment:
Can this entire else block be replaced by
```java
} else if (e.getEventType() == MouseEvent.MOUSE_CLICKED && e.getButton() == MouseButton.MIDDLE) {
	logger.debug("Middle button pressed {}x {}", e.getClickCount(), System.currentTimeMillis());
	// Here we toggle between the MOVE tool and any previously selected tool
	if (getSelectedTool() == PathTools.MOVE)
		setSelectedTool(previousTool);
	else
		setSelectedTool(PathTools.MOVE);
}
```
This behaves well for me - and is much simpler, avoiding any need to retain the last button that was pressed.

Reply:
Well, after all the head scratching, the simplest code (the one you proposed) behaves the way it's supposed to... No more issues with fast clicks, so I'm not exactly sure why I was seeing the issue in the first place.

Anyway, the last two pushes should resolve the issues you raised!

Cheers,
Egor